composer.json

```diff
@@ -55,44 +48,106 @@
+            "php -r \"array_map('unlink', glob('./example/*/pacts/*.json'));\" --",
+            "phpunit --no-coverage"
+        ],
+        "gen-lib": [
+            "protoc --php_out=example/protobuf-sync-message/library/src example/protobuf-sync-message/library/proto/area_calculator.proto",
+            "protoc --php_out=example/protobuf-async-message/library/src example/protobuf-async-message/library/proto/say_hello.proto"
+        ],
+        "check-compatibility": "behat"
     },
     "extra": {
         "downloads": {
-            "pact-ruby-standalone": {
-                "version": "2.4.1",
+            "pact-ffi-headers": {
+                "version": "0.4.18",
+                "url": "https://github.com/pact-foundation/pact-reference/releases/download/libpact_ffi-v{$version}/pact.h",
+                "path": "bin/pact-ffi-headers/pact.h"
+            },
+            "pact-ffi-lib": {
+                "version": "0.4.18",
+                "variables": {
+                    "{$prefix}": "PHP_OS_FAMILY === 'Windows' ? 'pact_ffi' : 'libpact_ffi'",
+                    "{$os}": "PHP_OS === 'Darwin' ? 'osx' : strtolower(PHP_OS_FAMILY)",
+                    "{$architecture}": "in_array(php_uname('m'), ['arm64', 'aarch64']) ? (PHP_OS === 'Darwin' ? 'aarch64-apple-darwin' : 'aarch64') : 'x86_64'",
+                    "{$extension}": "PHP_OS_FAMILY === 'Windows' ? 'dll' : (PHP_OS === 'Darwin' ? 'dylib' : 'so')"
+                },
+                "url": "https://github.com/pact-foundation/pact-reference/releases/download/libpact_ffi-v{$version}/{$prefix}-{$os}-{$architecture}.{$extension}.gz",
+                "path": "bin/pact-ffi-lib/pact.{$extension}"
+            },
+            "pact-stub-server": {
+                "version": "0.5.3",
                 "variables": {
-                    "{$os}": "PHP_OS_FAMILY === 'Windows' ? 'windows' : (PHP_OS === 'Darwin' ? 'osx' : 'linux')",
-                    "{$architecture}": "strtolower(php_uname('m')) === 'arm64' || strtolower(php_uname('m')) === 'aarch64' ? '-arm64' : (strtolower(php_uname('m')) === 'x86' && PHP_OS_FAMILY === 'Windows' ? '-x86' : '-x86_64')",
-                    "{$extension}": "PHP_OS_FAMILY === 'Windows' ? 'zip' : 'tar.gz'"
+                    "{$os}": "PHP_OS === 'Darwin' ? 'osx' : strtolower(PHP_OS_FAMILY)",
+                    "{$architecture}": "in_array(php_uname('m'), ['arm64', 'aarch64']) ? 'aarch64' : 'x86_64'",
+                    "{$extension}": "PHP_OS_FAMILY === 'Windows' ? '.exe' : ''"
                 },
-                "url": "https://github.com/pact-foundation/pact-ruby-standalone/releases/download/v{$version}/pact-{$version}-{$os}{$architecture}.{$extension}",
-                "path": "bin/pact-ruby-standalone"
+                "url": "https://github.com/pact-foundation/pact-stub-server/releases/download/v{$version}/pact-stub-server-{$os}-{$architecture}{$extension}.gz",
+                "path": "bin/pact-stub-server/pact-stub-server{$extension}",
```

Comment:
very clever; I don't see away around having something like this, but... Risk++

This affects BOM as well as clarity, and would be outside of any lockfile.

Reply:
So this comes up often and it's seems like its been a tough nut to crack over the course of the Pact project.

Pact isn't usable without the shared libs (whether ruby or rust, due to a shared core)

Some client languages need to provide additional binding files, aswell as the shared core.

The shared core is built for different platforms and architectures.

Some languages do not easily provide a mechanism to satisfactorily provide platform/arch specific packages, in a single meta package (Python does for, JavaScript doesn't)

So you end up in a case where you have

- A single package with every platform/arch combo of shared library in
  - easiest for users, everything in one place
  - big fat bundles with wastage, if users never run on certain platforms
  - overkill on package managers that can support platform/arch specific combos
- Multiple packages, one for each platform/arch combo
  - Ideal if package manager can detect
  - Less than ideal if package manager cannot detect, and onus is on the user to select the right package, for the selected platform/arch. This can be difficult if you run a project on multiple platform/archs. Needs lots of documentation
- Really slim package that downloads runtime libraries for specified platform/combo
  - Smallest package size possible
  - Requires external files to be useful (needs the shared core)
  - big issue for users with corp proxies, no external access
  - needs ability for users to provide their own binaries and skip downloading of files
  
The shared libraries can be highly trusted as they are generated from CI and pulled from releases, they can be shasum checked, as sha's as also published. The sha's could be stored in the released client libraries packages such that if an attacker updates the pact ref releases, and the associated sha, the client library package would reject it, as it stores it's own copy for of the shasum
  
I do wonder how much users worry about the size, for me, it is a real concern as I run multiple example projects so have huge numbers of duplication, of binaries/executables that can be made smaller (something else I am looking at)

It may be also useful to consider where practical for the shared library to be installed to a common location such as a users `$HOME/.pact` library, and client libraries looking there. That way we could have an installer that can deal with getting you a known shared library, and the pact client libraries don't have to concern with packaging it, just referencing it by version in a known location.

Surfacing the version used of the client libraries easily, in vanilla client language without requiring ffi func or extensions gives users the best chance of being able to find out the version of the pact client libs they need, so they can install them.

I'd like alot of these concerns really to be taken up with a pact cli tool that can install/audit and verify the shared libs, and a doctor command that can check an env/system's suitability for each client lang

Reply:
> The shared libraries can be highly trusted as they are generated from CI and pulled from releases, they can be shasum checked, as sha's as also published. The sha's could be stored in the released client libraries packages such that if an attacker updates the pact ref releases, and the associated sha, the client library package would reject it, as it stores it's own copy for of the shasum

I think it's a great idea, and we can add this feature to https://github.com/tienvx/composer-downloads-plugin (or https://github.com/pact-foundation/composer-downloads-plugin if we made it available first)

Reply:
So I don't disagree with anything said here, and I don't think anyone should feel their position is invalid.

Tien, your suggestion looks like it would mitigate a lot of the risk.

What exists works. The continued call-out are improvement suggestions. The more objections the project can overcome, the easier adoption should become. PACT is a heavyweight implementation to existing non-greenfield, although the barrier for greenfield is low.

Reply:
@tienvx thanks for continuing to put up with my suggestions etc. Now this is within `pact-foundation` I think the sha checking can be a separate future-thing

Reply:
> is there anyone that can help get a packagist package published for that repo now it's within pact-foundation?

Let me ask in Slack Maintainer channel.

Reply:
Look like he is very busy. We can use `repositories` feature of Composer to allow installing the package for now  https://github.com/pact-foundation/pact-php/pull/544